builder/format.go

```diff
@@ -0,0 +1,129 @@
+package builder
+
+import (
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+	"strconv"
+	"strings"
+
+	"github.com/clearlinux/mixer-tools/helpers"
+	"github.com/pkg/errors"
+)
+
+// UpdateFormatVersion updates the builder.conf file with a new format version
+func UpdateFormatVersion(b *Builder, version string) error {
+	b.Config.Swupd.Format = version
+
+	newver := "${1}" + b.Config.Swupd.Format
+	var re = regexp.MustCompile(`(FORMAT=)[0-9]*`)
+
+	builderData, err := ioutil.ReadFile(b.BuildConf)
+	if err != nil {
+		return errors.Wrap(err, "Failed to read builder.conf")
+	}
+
+	builderEdit := re.ReplaceAllString(string(builderData), newver)
+
+	var filename string
+	if UseNewConfig {
+		filename, err = GetConfigPath("")
+		if err != nil {
+			return err
+		}
+
+		return b.Config.SaveConfig(filename)
+	}
+
+	builderOut := []byte(builderEdit)
+	if err = ioutil.WriteFile(b.BuildConf, builderOut, 0644); err != nil {
+		return errors.Wrap(err, "Failed to write new builder.conf")
+	}
+
+	return nil
+}
+
+// CopyFullGroupsINI copies the initial ini file which has ALL bundle definitions
+func CopyFullGroupsINI(b *Builder) error {
+	return helpers.CopyFile(filepath.Join(b.Config.Builder.ServerStateDir, "full_groups.ini"), filepath.Join(b.Config.Builder.ServerStateDir, "groups.ini"))
+}
+
+// CopyTrimmedGroupsINI copies the new ini made with deleted bundles removed
+func CopyTrimmedGroupsINI(b *Builder) error {
+	return helpers.CopyFile(filepath.Join(b.Config.Builder.ServerStateDir, "trimmed_groups.ini"), filepath.Join(b.Config.Builder.ServerStateDir, "groups.ini"))
+}
+
+// RevertFullGroupsINI copies back the full ini to the manifest creator accounts for deleted bundles
+func RevertFullGroupsINI(b *Builder) error {
+	return helpers.CopyFile(filepath.Join(b.Config.Builder.ServerStateDir, "groups.ini"), filepath.Join(b.Config.Builder.ServerStateDir, "full_groups.ini"))
+}
+
+// RevertTrimmedGroupsINI copies back the trimmed INI so manifests are not created anymore for deleted bundles in new format
+func RevertTrimmedGroupsINI(b *Builder) error {
+	return helpers.CopyFile(filepath.Join(b.Config.Builder.ServerStateDir, "groups.ini"), filepath.Join(b.Config.Builder.ServerStateDir, "trimmed_groups.ini"))
+}
+
+func getLastBuildVersion(b *Builder) (string, error) {
+	var lastVer []byte
+	var err error
+
+	filename := filepath.Join(b.Config.Builder.ServerStateDir, "image/LAST_VER")
+	// Likely the first build or some weirdness that requires us to get the container image and build it anyway
+	if lastVer, err = ioutil.ReadFile(filename); err != nil {
+		return "", errors.Wrap(err, "Cannot find last built version, must get container image")
+	}
+	data := string(lastVer)
+	ver := strings.Split(data, "\n")
+
+	return ver[0], nil
+}
+
+// CheckBumpNeeded returns nil if it successfully deduces there is no format
+// bump boundary being crossed.
+func CheckBumpNeeded(b *Builder) (bool, error) {
+	version, err := getLastBuildVersion(b)
+	if err != nil {
+		return false, err
```

Comment:
This needs to just `return false, nil` here, not `err`, because a first-time build isn't a failure case. It's theoretically possible that something else would cause an error here (the file not being readable??), but if this returns an error, whatever code calls `CheckBumpNeeded` will fail on first time builds.

Reply:
I think the code should check whether it returns true or false, not err, to decide to do a bump. If it returns false, err -> then no bump is needed but we need to get a container regardless.